lynxkite-graph-analytics/src/lynxkite_graph_analytics/pykeen_ops.py

```diff
@@ -0,0 +1,633 @@
+    if relation_embeddings is None:
+        raise AttributeError(
+            f"Cannot extract relation embeddings from model type: {type(actual_model)}. "
+            f"Available attributes: {[attr for attr in dir(actual_model) if not attr.startswith('_')]}"
+        )
+
+    print(relation_embeddings.shape)
+
+    relations_table = bundle.dfs["relations"]
+    if "embedding" not in relations_table.columns:
+        relations_table["embedding"] = None
+    relations_table["embedding"] = pd.Series(
+        [relation_embeddings[relation_labels.index(label)].numpy() for label in relation_labels]
+    )
+    bundle.dfs["relations"] = relations_table
+
+    return bundle
+
+
+class EvaluatorTypes(str, enum.Enum):
+    ClassificationEvaluator = "Classification Evaluator"
+    MacroRankBasedEvaluator = "Macro Rank Based Evaluator"
+    RankBasedEvaluator = "Rank Based Evaluator"
+
+    def to_class(self) -> evaluation.Evaluator:
+        return getattr(evaluation, self.name.replace(" ", ""))()
+
+
+@op("Evaluate model")
+def evaluate(
+    bundle: core.Bundle,
+    *,
+    model_name: core.PyKEENModelName = "PyKEENmodel",
+    evaluator_type: EvaluatorTypes = EvaluatorTypes.RankBasedEvaluator,
+    metrics_str: str = "ALL",
+):
+    """Metrics are a comma separated list, "ALL" if all metrics are needed"""
+    bundle = bundle.copy()
+    model: PyKEENModelWrapper = bundle.other.get(model_name)
+    entity_to_id = model.entity_to_id
+    relation_to_id = model.relation_to_id
+    evaluator = evaluator_type.to_class()
+    testing_triples = TriplesFactory.from_labeled_triples(
+        bundle.dfs["edges_test"][["head", "relation", "tail"]].to_numpy(),
+        entity_to_id=entity_to_id,
+        relation_to_id=relation_to_id,
+    )
+    training_triples = TriplesFactory.from_labeled_triples(
+        bundle.dfs["edges_train"][["head", "relation", "tail"]].to_numpy(),
+        entity_to_id=entity_to_id,
+        relation_to_id=relation_to_id,
+    )
+    validation_triples = TriplesFactory.from_labeled_triples(
+        bundle.dfs["edges_val"][["head", "relation", "tail"]].to_numpy(),
+        entity_to_id=entity_to_id,
+        relation_to_id=relation_to_id,
+    )
+
+    evaluated = evaluator.evaluate(
+        model=model.model,
+        mapped_triples=testing_triples.mapped_triples,
+        additional_filter_triples=[
+            training_triples.mapped_triples,
+            validation_triples.mapped_triples,
+        ],
+        batch_size=32,
+    )
+    if metrics_str == "ALL":
+        bundle.dfs["metrics"] = evaluated.to_df()
+        return bundle
+
+    metrics = metrics_str.split(",")
+    metrics_df = pd.DataFrame(columns=["metric", "score"])
+
+    for metric in metrics:
+        metric = metric.strip()
+        try:
+            score = evaluated.get_metric(metric)
+        except Exception as e:
+            raise Exception(f"Possibly unknown metric: {metric}\nError: {e}")
```

Comment:
Use [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining):

```suggestion
            raise Exception(f"Possibly unknown metric: {metric}") from e
```